### DISCUSSION TITLE
Click on an element by index

### DISCUSSION THREAD

**Discussion by ianrhamilton** [AdevintaSpain/Barista#389] [Q&A] [answered]
Hi, I have a view that contains one to many views with the same ids, and with barista, I cannot figure out a way of accessing a particular element by its index. For example, my screen contains three boarding passes that I swipe to check elements on these cards.  

The only workaround that I have found so far is to create my own `clickOn()` function: 

```

 fun clickOn(id: Int, index: Int) {
        findWithIndex(id, index).perform(click())
    }

private fun findWithIndex(viewId: Int, index: Int): ViewInteraction {
        return onView(withIndex(withId(viewId), index))
    }

private fun withIndex(matcher: Matcher<View?>, index: Int): Matcher<View?>? {
        return object : TypeSafeMatcher<View?>() {
            var currentIndex = 0
            override fun describeTo(description: Description) {
                description.appendText("with index: ")
                description.appendValue(index)
                matcher.describeTo(description)
            }

            override fun matchesSafely(view: View?): Boolean {
                return matcher.matches(view) && currentIndex++ == index
            }
        }
    }
```

This does serve my purpose, however, I wondered if there was something I was missing with Barista that I could leverage? Otherwise, i'd have to create functions for  `clickOn` and `getText()` etc.

Thanks,
Ian

**Comment by Sloy** [comment, marked as answer]
Hello @ianrhamilton!
Do your items contain different texts? Do you know their text?
In that case, you could use `clickOn("Boarding pass 1")` instead of the view id.

**Reply by ianrhamilton** [reply]
hey @Sloy! Thanks for the speedy response. That's a good question! I think my approach was wrong. What I was trying to achieve was swiping right on an element (that has duplicate id's), and then getting the text of elements.

I was able to achieved this successfully by doing:
`swipeViewPagerForward()` and `assertDisplayed("Boarding pass 1")` (as I am using mocked data the text is always the same). Thanks for your suggestion :) 

Great work on Barists by the way, I love it. I've started refactoring our espresso tests to use Barista now.

**Reply by Sloy** [reply]
Great! I'm glad I could help and you found a solution :) 

Good luck with the refactoring! I hope you can simplify your code with Barista :D

**Reply by anefedov** [reply]
Hi @Sloy 
Similar question, but in my case, we have buttons without text, so we only can use the index. 
`clickOn()` only accepts Int or String. What would be your recommendation?

**Reply by Sloy** [reply]
Hi @anefedov. Your options depend. If your buttons are on a ListView or RecyclerView, there are specialized functions for that. Otherwise, if there is only one button visible at a time, `clickOn()` should find the one visible. If not, you are out of luck. I don't think we have a function for that in Barista. You should write some custom Espresso code for that. Something like `onView(firstOf(withText("My button")).perform(onClick())`. 
(You should find some matcher like `firstOf` by Googling, I don't think we have one in Barista but I can't check right now)

**Reply by ianrhamilton** [reply]
> Hi @Sloy Similar question, but in my case, we have buttons without text, so we only can use the index. `clickOn()` only accepts Int or String. What would be your recommendation?

You can use the workaround I provided above for clicking an element by index.
